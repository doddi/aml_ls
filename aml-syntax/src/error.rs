use thiserror::Error;

pub type Result<T, E = Error> = std::result::Result<T, E>;

#[derive(Debug, Clone, Error)]
pub enum Error {
    #[error("Parse error")]
    Parse,
}