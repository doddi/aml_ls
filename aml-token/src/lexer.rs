use std::iter::Peekable;
use std::str::CharIndices;

use crate::error::Result;
use crate::token::{Element, IntoToken, LexError, Operator, Primitive, Token, TokenKind};

pub struct Lexer<'lex> {
    chars: Peekable<CharIndices<'lex>>,
    content: &'lex str,
}

impl<'lex> Lexer<'lex> {
    pub fn new(content: &'lex str) -> Self {
        Self {
            chars: content.char_indices().peekable(),
            content,
        }
    }

    fn next_token(&mut self) -> Result<Token> {
        let (index, curr) = match self.chars.next() {
            None => return self.eof(),
            Some(curr) => curr,
        };

        let next = self.chars.peek().map(|(_, c)| *c);

        match (curr, next) {
            ('/', Some('/')) => {
                self.chars.next(); // consume the second slash
                loop {
                    if let Some((_, '\n')) | None = self.chars.peek() {
                        break;
                    };
                    self.chars.next();
                }
                self.next_token()
            }
            ('&', Some('&')) => {
                let _ = self.chars.next();
                Ok(Operator::And.into_token(index, index + 2))
            }
            ('|', Some('|')) => {
                let _ = self.chars.next();
                Ok(Operator::Or.into_token(index, index + 2))
            }
            ('=', Some('=')) => {
                let _ = self.chars.next();
                Ok(Operator::EqualEqual.into_token(index, index + 2))
            }
            ('!', Some('=')) => {
                let _ = self.chars.next();
                Ok(Operator::NotEqual.into_token(index, index + 2))
            }
            ('>', Some('=')) => {
                let _ = self.chars.next();
                Ok(Operator::GreaterThanOrEqual.into_token(index, index + 2))
            }
            ('<', Some('=')) => {
                let _ = self.chars.next();
                Ok(Operator::LessThanOrEqual.into_token(index, index + 2))
            }
            ('-', Some('>')) => {
                let _ = self.chars.next();
                Ok(Operator::Association.into_token(index, index + 2))
            }

            ('(', _) => Ok(Operator::LParen.into_token(index, index + 1)),
            (')', _) => Ok(Operator::RParen.into_token(index, index + 1)),
            ('[', _) => Ok(Operator::LBracket.into_token(index, index + 1)),
            (']', _) => Ok(Operator::RBracket.into_token(index, index + 1)),
            ('{', _) => Ok(Operator::LCurly.into_token(index, index + 1)),
            ('}', _) => Ok(Operator::RCurly.into_token(index, index + 1)),
            (':', _) => Ok(Operator::Colon.into_token(index, index + 1)),
            (',', _) => Ok(Operator::Comma.into_token(index, index + 1)),
            ('.', _) => Ok(Operator::Dot.into_token(index, index + 1)),
            ('!', _) => Ok(Operator::Not.into_token(index, index + 1)),
            ('+', _) => Ok(Operator::Plus.into_token(index, index + 1)),
            ('-', _) => Ok(Operator::Minus.into_token(index, index + 1)),
            ('*', _) => Ok(Operator::Mul.into_token(index, index + 1)),
            ('/', _) => Ok(Operator::Div.into_token(index, index + 1)),
            ('%', _) => Ok(Operator::Mod.into_token(index, index + 1)),
            ('>', _) => Ok(Operator::GreaterThan.into_token(index, index + 1)),
            ('<', _) => Ok(Operator::LessThan.into_token(index, index + 1)),
            ('?', _) => Ok(Operator::Either.into_token(index, index + 1)),
            ('=', _) => Ok(TokenKind::Equal.into_token(index, index + 1)),
            ('\n', _) => Ok(TokenKind::Newline.into_token(index, index + 1)),
            ('@', _) => Ok(TokenKind::Component.into_token(index, index + 1)),
            ('$', _) => Ok(TokenKind::ComponentSlot.into_token(index, index + 1)),

            ('a'..='z' | 'A'..='Z' | '_', _) => Ok(self.lex_identifier(index)),
            ('0'..='9', _) => Ok(self.lex_number(index)),
            ('"' | '\'', _) => Ok(self.lex_string(curr, index)),
            _ if curr.is_whitespace() => Ok(self.lex_indent(index)),
            ('#', Some('0'..='9' | 'a'..='f' | 'A'..='F')) => Ok(self.lex_hex_value(index)),
            _ => self.eof(),
        }
    }

    fn eof(&self) -> Result<Token> {
        let location = self.content.len()..self.content.len();
        Ok(Token(TokenKind::Eof, location.into()))
    }

    fn lex_identifier(&mut self, start_byte: usize) -> Token {
        let mut end_byte = start_byte;
        while let Some((e, 'a'..='z' | 'A'..='Z' | '_' | '|' | '0'..='9')) = self.chars.peek() {
            end_byte = *e;
            self.chars.next();
        }

        let str = &self.content[start_byte..=end_byte];

        let kind = match str {
            "for" => TokenKind::For,
            "in" => TokenKind::In,
            "if" => TokenKind::If,
            "else" => TokenKind::Else,
            "true" => TokenKind::Primitive(Primitive::Bool(true)),
            "false" => TokenKind::Primitive(Primitive::Bool(false)),
            "let" => TokenKind::Decl,
            "switch" => TokenKind::Switch,
            "case" => TokenKind::Case,
            "default" => TokenKind::Default,
            "with" => TokenKind::With,
            "as" => TokenKind::As,
            "text" => TokenKind::Element(Element::Text),
            "span" => TokenKind::Element(Element::Span),
            "border" => TokenKind::Element(Element::Border),
            "alignment" => TokenKind::Element(Element::Alignment),
            "vstack" => TokenKind::Element(Element::VStack),
            "hstack" => TokenKind::Element(Element::HStack),
            "zstack" => TokenKind::Element(Element::ZStack),
            "row" => TokenKind::Element(Element::Row),
            "column" => TokenKind::Element(Element::Column),
            "expand" => TokenKind::Element(Element::Expand),
            "position" => TokenKind::Element(Element::Position),
            "spacer" => TokenKind::Element(Element::Spacer),
            "overflow" => TokenKind::Element(Element::Overflow),
            "padding" => TokenKind::Element(Element::Padding),
            "canvas" => TokenKind::Element(Element::Canvas),
            "container" => TokenKind::Element(Element::Container),
            _ => TokenKind::Identifier((start_byte, end_byte + 1).into()),
        };

        kind.into_token(start_byte, end_byte + 1)
    }

    fn lex_number(&mut self, start_byte: usize) -> Token {
        let mut end_byte = start_byte;
        while let Some((e, '0'..='9' | '.')) = self.chars.peek() {
            end_byte = *e;
            self.chars.next();
        }

        let mut literal = &self.content[start_byte..=end_byte];
        let mut dots = literal.splitn(3, '.');

        if let (Some(one), Some(two), Some(_)) = (dots.next(), dots.next(), dots.next()) {
            // numeric literal contains two or more dots, so we only lex,
            // from the first numerical to the last numerical before the
            // second dot, <one.len()>.<two.len()>
            //
            // 123.456.789 -> 123.456 (leave .789) to be lexed after
            literal = &literal[..one.len() + 1 + two.len()];
        }

        let is_float = literal.contains('.');
        let kind = match is_float {
            true => {
                let literal = literal.parse().expect("failed to parse float literal");
                TokenKind::Primitive(Primitive::Float(literal))
            }
            false => {
                let literal = literal.parse().expect("failed to parse int literal");
                TokenKind::Primitive(Primitive::Int(literal))
            }
        };

        kind.into_token(start_byte, end_byte + 1)
    }

    fn lex_string(&mut self, delimiter: char, start_byte: usize) -> Token {
        loop {
            match self.chars.next() {
                Some((end_byte, next)) if next == delimiter => {
                    // add +1 to include the closing delimiter in the token
                    let end_byte = end_byte + 1;
                    let location = (start_byte, end_byte).into();
                    break TokenKind::String(location).into_token(start_byte, end_byte);
                }
                // Handle escape sequences
                Some((_, '\\')) => {
                    // found escape character, so consume the next character regardless of what it
                    // is. This properly handles \", \', \\, \n, etc.
                    if self.chars.next().is_none() {
                        // if there is no character, the string is unterminated
                        break LexError::UnterminatedString
                            .into_token(start_byte, self.content.len());
                    };
                }
                None => {
                    break LexError::UnterminatedString.into_token(start_byte, self.content.len());
                }
                _ => {}
            }
        }
    }

    fn lex_indent(&mut self, start_byte: usize) -> Token {
        let mut count = 1;

        loop {
            match self.chars.peek() {
                Some((_, next)) if next.is_whitespace() && *next != '\n' => {
                    count += 1;
                    self.chars.next();
                }
                Some(_) | None => break,
            }
        }

        TokenKind::Indent(count).into_token(start_byte, start_byte + count)
    }

    fn lex_hex_value(&mut self, start_byte: usize) -> Token {
        const SHORT: usize = 3;
        const LONG: usize = 6;

        let start_byte = start_byte + 1; // consume #
        let mut end_byte = start_byte;

        while let Some((_, '0'..='9' | 'a'..='f' | 'A'..='F')) = self.chars.peek() {
            self.chars.next();
            end_byte += 1;
        }

        let hex = &self.content[start_byte..end_byte];
        let len = hex.len();
        // Make sure that it's either three or six characters,
        // otherwise it's an invalid hex
        if len != 3 && len != 6 {
            return LexError::InvalidHex.into_token(start_byte, end_byte);
        }

        let kind = match len {
            SHORT => {
                let r = u8::from_str_radix(&hex[0..1], 16).expect("already parsed");
                let r = r << 4 | r;
                let g = u8::from_str_radix(&hex[1..2], 16).expect("already parsed");
                let g = g << 4 | g;
                let b = u8::from_str_radix(&hex[2..3], 16).expect("already parsed");
                let b = b << 4 | b;
                TokenKind::Primitive(Primitive::Hex((r, g, b).into()))
            }
            LONG => {
                let r = u8::from_str_radix(&hex[0..2], 16).expect("already parsed");
                let g = u8::from_str_radix(&hex[2..4], 16).expect("already parsed");
                let b = u8::from_str_radix(&hex[4..6], 16).expect("already parsed");
                TokenKind::Primitive(Primitive::Hex((r, g, b).into()))
            }
            _ => unreachable!(),
        };

        kind.into_token(start_byte, end_byte)
    }
}

impl Iterator for Lexer<'_> {
    type Item = Result<Token>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.next_token() {
            Ok(Token(TokenKind::Eof, _)) => None,
            result => Some(result),
        }
    }
}

#[cfg(test)]
mod tests {
    use aml_core::Location;
    use serde::Serialize;

    use super::*;

    #[derive(Debug, Serialize)]
    struct SnapshotToken<'tok> {
        kind: TokenKind,
        location: Location,
        value: &'tok str,
    }

    impl<'tok> SnapshotToken<'tok> {
        fn from_token(token: Token, content: &'tok str) -> Self {
            let value = &content[token.1.to_range()];

            Self {
                kind: token.0,
                location: token.1,
                value,
            }
        }
    }

    #[test]
    fn test_indentation() {
        let template = r#"
vstack        // 0 spaces
 text "hi"    // 1 space → Indent
   span       // 3 spaces → Indent
 text "bye"   // 1 space → Dedent
"#;

        let tokens = Lexer::new(template)
            .map(|t| t.unwrap())
            .map(|t| SnapshotToken::from_token(t, template))
            .collect::<Vec<_>>();

        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_attributes() {
        let template = r#"
vstack [width: 10, height: 3]
    text [foreground: "red"] "Hello you"
    border [height: -10]
"#;
        let tokens = Lexer::new(template)
            .map(|t| t.unwrap())
            .map(|t| SnapshotToken::from_token(t, template))
            .collect::<Vec<_>>();

        insta::assert_yaml_snapshot!(tokens);
    }
}

