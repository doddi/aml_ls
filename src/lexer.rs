use std::cmp::Ordering;
use std::ops::{Range, RangeBounds};
use std::str::FromStr;

use crate::error::{Error, Result};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Element {
    Text,
    Span,
    Border,
    Alignment,
    VStack,
    HStack,
    ZStack,
    Row,
    Column,
    Expand,
    Position,
    Spacer,
    Overflow,
    Padding,
    Canvas,
    Container,
}

impl FromStr for Element {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "text" => Ok(Self::Text),
            "span" => Ok(Self::Span),
            "border" => Ok(Self::Border),
            "alignment" => Ok(Self::Alignment),
            "vstack" => Ok(Self::VStack),
            "hstack" => Ok(Self::HStack),
            "zstack" => Ok(Self::ZStack),
            "row" => Ok(Self::Row),
            "column" => Ok(Self::Column),
            "expand" => Ok(Self::Expand),
            "position" => Ok(Self::Position),
            "spacer" => Ok(Self::Spacer),
            "overflow" => Ok(Self::Overflow),
            "padding" => Ok(Self::Padding),
            "canvas" => Ok(Self::Canvas),
            "container" => Ok(Self::Container),
            _ => Err("Nuh uh".into()),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Keyword {
    For,
    In,
    If,
    Else,
    Let,
    Switch,
    Case,
    Default,
    With,
    As,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum TokenKind {
    Element(Element),
    Keyword(Keyword),
    Identifier,
    String,
    Newline,
    Indent,
    Dedent,
}

impl IntoToken for TokenKind {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token {
        Token::new(self, (start_byte, end_byte).into())
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Location {
    pub start_byte: usize,
    pub end_byte: usize,
}

impl Location {
    pub fn new(start: usize, end: usize) -> Self {
        Self {
            start_byte: start,
            end_byte: end,
        }
    }
}
impl From<(usize, usize)> for Location {
    fn from((start_byte, end_byte): (usize, usize)) -> Self {
        Self {
            start_byte,
            end_byte,
        }
    }
}

impl From<Range<usize>> for Location {
    fn from(range: Range<usize>) -> Self {
        let start = match range.start_bound() {
            std::ops::Bound::Included(start) => *start,
            std::ops::Bound::Excluded(start) => *start,
            std::ops::Bound::Unbounded => {
                panic!("can only construct a location from bounded ranges")
            }
        };

        let end = match range.end_bound() {
            std::ops::Bound::Included(end) => *end,
            std::ops::Bound::Excluded(end) => *end,
            std::ops::Bound::Unbounded => {
                panic!("can only construct a location from bounded ranges")
            }
        };

        Location {
            start_byte: start,
            end_byte: end,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Token {
    pub kind: TokenKind,
    pub location: Location,
}

impl Token {
    pub fn new(kind: TokenKind, location: Location) -> Self {
        Self { kind, location }
    }
}

pub struct Lexer<'lex> {
    cursor: usize,
    slice: &'lex str,
    content: &'lex str,
    current_indent: usize,
}

pub trait IntoToken {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token;
}

impl<'lex> Lexer<'lex> {
    pub fn new(content: &'lex str) -> Self {
        Self {
            slice: content,
            content,
            cursor: 0,
            current_indent: 0,
        }
    }

    fn advance_by(&mut self, amount: usize) {
        self.slice = &self.slice[amount..];
        self.cursor += amount;
    }

    fn make_token<T>(&mut self, tokenizable: T, size: usize) -> Token
    where
        T: IntoToken,
    {
        let start_byte = self.cursor;
        self.advance_by(size);
        tokenizable.into_token(start_byte, self.cursor)
    }

    fn lex_string(&mut self) -> Token {
        let start_byte = self.cursor;
        self.advance_by(1);
        let end_of_string = self.slice.find(['"', '\n']).unwrap_or(self.slice.len());

        self.cursor += end_of_string;
        self.slice = &self.slice[end_of_string + 1..];

        TokenKind::String.into_token(start_byte, self.cursor)
    }

    fn lex_identifier(&mut self) -> Token {
        let next_whitespace = self
            .slice
            .find(|c| !matches!(c, 'a'..='z' | 'A'..='Z' | '_' | '0'..='9'))
            .unwrap_or(self.slice.len());

        let identifier = &self.slice[..next_whitespace];

        match Element::from_str(identifier) {
            Ok(element) => self.make_token(TokenKind::Element(element), next_whitespace),
            Err(_) => self.make_token(TokenKind::Identifier, next_whitespace),
        }
    }

    fn handle_indentation(&mut self, new_indent: usize) -> Option<Token> {
        let current_indent = self.current_indent;
        self.current_indent = new_indent;

        match new_indent.cmp(&current_indent) {
            Ordering::Greater => Some(TokenKind::Indent.into_token(self.cursor, self.cursor)),
            Ordering::Less => Some(TokenKind::Dedent.into_token(self.cursor, self.cursor)),
            Ordering::Equal => None,
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = Result<Token>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            let mut chars = self.slice.chars().peekable();
            let curr = chars.next()?;
            let next = chars.peek();

            break match (curr, next) {
                ('\n', _) => {
                    let start_byte = self.cursor;
                    self.advance_by(1);

                    let leading_spaces = self.slice.chars().take_while(|&c| c == ' ').count();
                    self.advance_by(leading_spaces);

                    let newline_token = TokenKind::Newline.into_token(start_byte, start_byte + 1);

                    match self.handle_indentation(leading_spaces) {
                        Some(indent_token) => Some(Ok(indent_token)),
                        None => Some(Ok(newline_token)),
                    }
                }
                (c, _) if c.is_whitespace() => {
                    self.advance_by(c.len_utf8());
                    continue;
                }

                ('/', Some('/')) => {
                    let eol_location = self.slice.find('\n').unwrap_or(self.slice.len());
                    self.advance_by(eol_location);
                    continue;
                }

                ('"', _) => Some(Ok(self.lex_string())),

                ('a'..='z' | 'A'..='Z' | '_', _) => Some(Ok(self.lex_identifier())),
                _ => todo!(),
            };
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lexer() {
        let template = r#"
vstack        // 0 spaces
 text "hi"    // 1 space → Indent
   span       // 3 spaces → Indent
 text "bye"   // 1 space → Dedent
"#;

        let tokens = Lexer::new(template).map(|t| t.unwrap()).collect::<Vec<_>>();

        println!("{tokens:#?}");
        panic!();
    }
}
