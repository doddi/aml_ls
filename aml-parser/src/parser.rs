use std::collections::HashMap;

use serde::Serialize;

use crate::error::Result;
use crate::lexer::{Element, Lexer, Location, Operator, TokenKind, TransposeRef};

#[macro_export]
macro_rules! expect {
    ($lexer:expr, $kind:pat) => {{
        let Some(token) = $lexer.next().transpose()? else {
            todo!();
        };

        if !matches!(token.kind, $kind) {
            todo!();
        }

        token
    }};
}

#[macro_export]
macro_rules! peek {
    ($lexer:expr) => {
        $lexer.peek().transpose().map_err(|e| e.clone())?
    };
}

#[macro_export]
macro_rules! expect_peek {
    ($lexer:expr, $kind:pat) => {{
        let token = peek!($lexer);
        if token.is_none() || !matches!(token.unwrap().kind, $kind) {
            todo!();
        }
        token.unwrap()
    }};
}

#[macro_export]
macro_rules! consume {
    ($lexer:expr) => {
        $lexer.next().transpose()?
    };
}

pub type NodeId = usize;

#[derive(Debug, Default)]
pub struct Ast {
    pub nodes: Vec<AstNode>,
    pub variables: HashMap<String, Location>,
    pub scopes: Vec<Scope>,
}

#[derive(Debug)]
pub enum AstNode {
    String {
        value: Location,
    },
    Text {
        value: Option<Box<AstNode>>,
        attributes: Vec<AstNode>,
        children: Vec<AstNode>,
        location: Location,
    },
    Span {
        value: Option<Box<AstNode>>,
        attributes: Vec<AstNode>,
        location: Location,
    },
    Identifier {
        value: Location,
    },
    Attribute {
        name: Box<AstNode>,
        value: Box<AstNode>,
    },
}

#[derive(Debug, Serialize)]
pub struct Scope {
    pub variables: Vec<String>,
    pub parent: Option<usize>,
}

pub struct Parser<'p> {
    lexer: Lexer<'p>,
    ast: Ast,
    scope_stack: Vec<usize>,
}

impl<'p> Parser<'p> {
    pub fn new(lexer: Lexer<'p>) -> Self {
        Self {
            lexer,
            ast: Ast::default(),
            scope_stack: Vec::new(),
        }
    }

    pub fn parse(mut self) -> Result<Ast> {
        self.add_scope()?;

        while !self.lexer.is_empty() {
            let Some(node) = self.parse_node()? else { continue };
            self.ast.nodes.push(node);
        }

        Ok(self.ast)
    }

    fn parse_node(&mut self) -> Result<Option<AstNode>> {
        let node = match peek!(self.lexer) {
            Some(token) if matches!(token.kind, TokenKind::Element(_)) => {
                Some(self.parse_element()?)
            }
            Some(token) if matches!(token.kind, TokenKind::String) => Some(self.parse_string()?),
            Some(token) if matches!(token.kind, TokenKind::Newline) => {
                consume!(self.lexer);
                None
            }
            t => todo!("{t:#?}"),
        };

        Ok(node)
    }

    fn parse_string(&mut self) -> Result<AstNode> {
        let token = expect!(self.lexer, TokenKind::String);
        let value = token.location;
        Ok(AstNode::String { value })
    }

    fn parse_element(&mut self) -> Result<AstNode> {
        let token = expect_peek!(self.lexer, TokenKind::Element(_));

        let TokenKind::Element(element) = token.kind else {
            unreachable!();
        };

        let node = match element {
            Element::Text => self.parse_text()?,
            Element::Span => self.parse_span()?,
            t => todo!("{t:?}"),
        };

        Ok(node)
    }

    fn parse_text(&mut self) -> Result<AstNode> {
        let keyword = expect!(self.lexer, TokenKind::Element(Element::Text));
        let location = keyword.location;

        let attributes = match peek!(self.lexer) {
            Some(token) if matches!(token.kind, TokenKind::Operator(Operator::LBracket)) => {
                self.parse_attributes()?
            }
            _ => vec![],
        };

        let value = match peek!(self.lexer) {
            Some(token) if matches!(token.kind, TokenKind::String) => {
                Some(Box::new(self.parse_string()?))
            }
            _ => None,
        };

        let mut children = vec![];

        while let Some(token) = peek!(self.lexer) {
            match token.kind {
                TokenKind::Indent => {
                    consume!(self.lexer);
                    self.add_scope()?;
                }
                TokenKind::Dedent => {
                    consume!(self.lexer);
                    self.pop_scope()?
                }
                TokenKind::Newline => {
                    consume!(self.lexer);
                    continue;
                }
                _ => {
                    let Some(node) = self.parse_node()? else { continue };
                    children.push(node)
                }
            }
        }

        Ok(AstNode::Text {
            value,
            attributes,
            children,
            location,
        })
    }

    fn parse_identifier(&mut self) -> Result<AstNode> {
        let identifier = expect!(self.lexer, TokenKind::Identifier);

        Ok(AstNode::Identifier {
            value: identifier.location,
        })
    }

    fn parse_attributes(&mut self) -> Result<Vec<AstNode>> {
        expect!(self.lexer, TokenKind::Operator(Operator::LBracket));

        let mut attributes = vec![];

        loop {
            match peek!(self.lexer) {
                Some(token) if matches!(token.kind, TokenKind::Operator(Operator::RBracket)) => {
                    consume!(self.lexer);
                    break;
                }
                None => break,
                _ => {}
            }

            let name = self.parse_identifier()?;
            expect!(self.lexer, TokenKind::Operator(Operator::Colon));
            let value = self.parse_expression()?;

            attributes.push(AstNode::Attribute {
                name: Box::new(name),
                value: Box::new(value),
            })
        }

        Ok(attributes)
    }

    fn parse_expression(&mut self) -> Result<AstNode> {
        let lhs = match peek!(self.lexer) {
            Some(token) if matches!(token.kind, TokenKind::String) => self.parse_string()?,
            _ => todo!(),
        };

        Ok(lhs)
    }

    fn parse_span(&mut self) -> Result<AstNode> {
        let keyword = expect!(self.lexer, TokenKind::Element(Element::Span));
        let location = keyword.location;

        let attributes = match peek!(self.lexer) {
            Some(token) if matches!(token.kind, TokenKind::Operator(Operator::LBracket)) => {
                self.parse_attributes()?
            }
            _ => vec![],
        };

        let value = match peek!(self.lexer) {
            Some(token) if matches!(token.kind, TokenKind::String) => {
                Some(Box::new(self.parse_string()?))
            }
            _ => None,
        };

        Ok(AstNode::Span {
            value,
            attributes,
            location,
        })
    }

    fn add_scope(&mut self) -> Result<()> {
        let scope_id = self.ast.scopes.len();
        let parent = self.scope_stack.last().copied();

        self.ast.scopes.push(Scope {
            variables: Vec::new(),
            parent,
        });

        self.scope_stack.push(scope_id);
        Ok(())
    }

    fn pop_scope(&mut self) -> Result<()> {
        self.scope_stack.pop();
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;

    #[derive(Serialize)]
    struct SnapshotAst<'ast> {
        pub nodes: Vec<SnapshotAstNode<'ast>>,
        pub variables: HashMap<String, Location>,
        pub scopes: Vec<Scope>,
    }

    impl<'ast> SnapshotAst<'ast> {
        pub fn from_ast(ast: Ast, content: &'ast str) -> Self {
            Self {
                nodes: ast
                    .nodes
                    .into_iter()
                    .map(|n| SnapshotAstNode::from_node(n, content))
                    .collect(),
                variables: ast.variables,
                scopes: ast.scopes,
            }
        }
    }

    #[derive(Serialize)]
    enum SnapshotAstNode<'ast> {
        String {
            value: Location,
            text: &'ast str,
        },
        Text {
            value: Option<Box<SnapshotAstNode<'ast>>>,
            attributes: Vec<SnapshotAstNode<'ast>>,
            children: Vec<SnapshotAstNode<'ast>>,
            location: Location,
            text: &'ast str,
        },
        Span {
            value: Option<Box<SnapshotAstNode<'ast>>>,
            attributes: Vec<SnapshotAstNode<'ast>>,
            location: Location,
            text: &'ast str,
        },
        Identifier {
            value: Location,
            text: &'ast str,
        },
        Attribute {
            value: Box<SnapshotAstNode<'ast>>,
            name: Box<SnapshotAstNode<'ast>>,
        },
    }

    impl<'ast> SnapshotAstNode<'ast> {
        fn from_node(node: AstNode, content: &'ast str) -> Self {
            match node {
                AstNode::String { value } => {
                    let text = &content[value.to_range()];
                    Self::String { value, text }
                }
                AstNode::Text {
                    value,
                    children,
                    location,
                    attributes,
                } => Self::Text {
                    value: value.map(|n| Box::new(SnapshotAstNode::from_node(*n, content))),
                    text: &content[location.to_range()],
                    attributes: attributes
                        .into_iter()
                        .map(|n| SnapshotAstNode::from_node(n, content))
                        .collect(),
                    children: children
                        .into_iter()
                        .map(|n| SnapshotAstNode::from_node(n, content))
                        .collect(),
                    location,
                },
                AstNode::Span {
                    value,
                    location,
                    attributes,
                } => Self::Span {
                    value: value.map(|n| Box::new(SnapshotAstNode::from_node(*n, content))),
                    attributes: attributes
                        .into_iter()
                        .map(|n| SnapshotAstNode::from_node(n, content))
                        .collect(),
                    text: &content[location.to_range()],
                    location,
                },
                AstNode::Identifier { value } => Self::Identifier {
                    text: &content[value.to_range()],
                    value,
                },
                AstNode::Attribute { name, value } => Self::Attribute {
                    name: Box::new(SnapshotAstNode::from_node(*name, content)),
                    value: Box::new(SnapshotAstNode::from_node(*value, content)),
                },
            }
        }
    }

    #[test]
    fn test_parser() {
        let template = r#"text "Hello"
    span "World""#;

        let lexer = Lexer::new(template);
        let parser = Parser::new(lexer);

        let ast = SnapshotAst::from_ast(parser.parse().unwrap(), template);

        insta::assert_yaml_snapshot!(ast);
    }

    #[test]
    fn test_parsing_attributes() {
        let template = r#"
text [foreground: "red"] "Hello"
    span [background: "yellow"] "World"
"#;

        let lexer = Lexer::new(template);
        let parser = Parser::new(lexer);

        let ast = SnapshotAst::from_ast(parser.parse().unwrap(), template);

        insta::assert_yaml_snapshot!(ast);
    }
}
