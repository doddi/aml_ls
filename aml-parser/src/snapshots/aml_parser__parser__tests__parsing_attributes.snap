---
source: aml-parser/src/parser.rs
expression: ast
---
nodes:
  - Text:
      value:
        String:
          value:
            start_byte: 26
            end_byte: 33
          text: "\"Hello\""
      attributes:
        - Attribute:
            value:
              String:
                value:
                  start_byte: 19
                  end_byte: 24
                text: "\"red\""
            name:
              Identifier:
                value:
                  start_byte: 7
                  end_byte: 17
                text: foreground
      children:
        - Span:
            value:
              String:
                value:
                  start_byte: 66
                  end_byte: 73
                text: "\"World\""
            attributes:
              - Attribute:
                  value:
                    String:
                      value:
                        start_byte: 56
                        end_byte: 64
                      text: "\"yellow\""
                  name:
                    Identifier:
                      value:
                        start_byte: 44
                        end_byte: 54
                      text: background
            location:
              start_byte: 38
              end_byte: 42
            text: span
      location:
        start_byte: 1
        end_byte: 5
      text: text
variables: {}
scopes:
  - variables: []
    parent: ~
  - variables: []
    parent: 0
